pkg/sandbox/sandbox.go

```diff
@@ -0,0 +1,306 @@
+
+	toolchainApiUrl, err := s.GetOpenshiftRouteHost(DEFAULT_TOOLCHAIN_NAMESPACE, DEFAULT_TOOLCHAIN_INSTANCE_NAME)
+	if err != nil {
+		return nil, err
+	}
+
+	if s.KeycloakUrl, err = s.GetOpenshiftRouteHost(DEFAULT_KEYCLOAK_NAMESPACE, DEFAULT_KEYCLOAK_INSTANCE_NAME); err != nil {
+		return nil, err
+	}
+
+	err = s.KubeRest.Get(context.Background(), types.NamespacedName{
+		Name:      userName,
+		Namespace: DEFAULT_TOOLCHAIN_NAMESPACE,
+	}, userSignup)
+	if err != nil {
+		if k8sErrors.IsNotFound(err) {
+			klog.Infof("user %s don't exists... recreating", userName)
+		}
+	} else {
+		userToken, err := s.GetKeycloakToken(DEFAULT_KEYCLOAK_TEST_CLIENT_ID, userName, userName, DEFAULT_KEYCLOAK_TESTING_REALM)
+		if err != nil {
+			return nil, err
+		}
+		if err := s.RegisterSandboxUser(userName); err != nil {
+			return nil, err
+		}
+
+		return s.GetKubeconfigPathForSpecifigUser(toolchainApiUrl, userName, kubeconfigPath, userToken)
+	}
+
+	if err := s.IsKeycloakRunning(); err != nil {
+		return nil, err
+	}
+
+	adminSecret, err := s.GetKeycloakAdminSecret()
+	if err != nil {
+		return nil, err
+	}
+
+	adminToken, err := s.GetKeycloakToken(DEFAULT_KEYCLOAK_ADMIN_CLIENT_ID, DEFAULT_KEYCLOAK_ADMIN_USERNAME, adminSecret, DEFAULT_KEYCLOAK_MASTER_REALM)
+	if err != nil {
+		return nil, err
+	}
+
+	if s.KeycloakUserExists(DEFAULT_KEYCLOAK_TESTING_REALM, adminToken.AccessToken, userName) {
+		registerUser, err := s.RegisterKeyclokUser(userName, adminToken.AccessToken, DEFAULT_KEYCLOAK_TESTING_REALM)
+		if err != nil && registerUser.Username == "" {
+			return nil, errors.New("failed to register user in keycloak: " + err.Error())
+		}
+	}
+
+	if err := s.RegisterSandboxUser(userName); err != nil {
+		return nil, err
+	}
+
+	userToken, err := s.GetKeycloakToken(DEFAULT_KEYCLOAK_TEST_CLIENT_ID, userName, userName, DEFAULT_KEYCLOAK_TESTING_REALM)
+	if err != nil {
+		return nil, err
+	}
+
+	return s.GetKubeconfigPathForSpecifigUser(toolchainApiUrl, userName, kubeconfigPath, userToken)
+}
+
+func (s *SandboxController) GetKubeconfigPathForSpecifigUser(toolchainApiUrl string, userName string, kubeconfigPath string, keycloakAuth *KeycloakAuth) (*SandboxUserAuthInfo, error) {
+	kubeconfig := api.NewConfig()
+	kubeconfig.Clusters[toolchainApiUrl] = &api.Cluster{
+		Server:                toolchainApiUrl,
+		InsecureSkipTLSVerify: true,
+	}
+	kubeconfig.Contexts[fmt.Sprintf("%s/%s/%s", userName, toolchainApiUrl, userName)] = &api.Context{
+		Cluster:   toolchainApiUrl,
+		Namespace: fmt.Sprintf("%s-tenant", userName),
+		AuthInfo:  fmt.Sprintf("%s/%s", userName, toolchainApiUrl),
+	}
+	kubeconfig.AuthInfos[fmt.Sprintf("%s/%s", userName, toolchainApiUrl)] = &api.AuthInfo{
+		Token: keycloakAuth.AccessToken,
+	}
+	kubeconfig.CurrentContext = fmt.Sprintf("%s/%s/%s", userName, toolchainApiUrl, userName)
+
+	err := clientcmd.WriteToFile(*kubeconfig, kubeconfigPath)
```

Comment:
just curious - why writing the kubeconfig file instead of creating a k8s client ?

Reply:
if u want to debug something with the user we want to return the kubeconfig

Reply:
Ok, now I see , it's saved as an "artifact" ! Thanks !